otherlibs/fiber/pool.ml

```diff
@@ -20,26 +16,58 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () = { tasks = Queue.create (); runner = None; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    match t.runner with
+    | None ->
+      Queue.push t.tasks f;
+      k ()
+    | Some r ->
+      assert (Queue.is_empty t.tasks);
```

Comment:
Could we put the task queue under the `None` branch to statically eliminate this assert?

That is, turn `runner` into something like

```ocaml
type runner = 
    | None of (unit -> unit t) Queue.t
    | Some of unit k
```

Reply:
(That adds an extra indirection in the `None` case but I'm unsure if that matters much.)

Reply:
Let me try. It would have also been nice to get rid of the boxing around `runner`. But I fear it would be a bit too magical.

Reply:
Okay, it doesn't really help. The continuation that calls `read` inside `run` must acquire `t` and the queue from the environment (so that we only allocate it once). Therefore, it must add pattern match on this new `runner` argument and then the `assert false` just moves there.

Reply:
OK. I think it's worth writing this down too to help the reader make sense of why the current implementation is like it is.

Reply:
Here's my attempt for reference:

```ocaml
diff --git a/otherlibs/fiber/fiber.mli b/otherlibs/fiber/fiber.mli
index 7bf3b131a..7eeee0273 100644
--- a/otherlibs/fiber/fiber.mli
+++ b/otherlibs/fiber/fiber.mli
@@ -388,7 +388,10 @@ module Pool : sig
       is called, [task pool ~f] will fail to submit new tasks.
 
       Note that stopping the pool does not prevent already queued tasks from
-      running. *)
+      running.
+
+      [stop pool] subsequent calls to [stop] ignored. That is to say, this
+      function is idempotent *)
   val stop : t -> unit fiber
 
   (** [run pool] Runs all tasks submitted to [pool] in parallel. Errors raised
diff --git a/otherlibs/fiber/pool.ml b/otherlibs/fiber/pool.ml
index da9f4d7c7..d3e93fb02 100644
--- a/otherlibs/fiber/pool.ml
+++ b/otherlibs/fiber/pool.ml
@@ -10,9 +10,13 @@ type status =
    It's implemented by a simple queue of thunks and a continuation to resume
    [run] whenever it runs out of work. *)
 
+type state =
+  | Stopped of unit
+  | Waiting_for_tasks of unit k
+  | Waiting_to_execute of (unit -> unit t) Queue.t
+
 type nonrec t =
-  { tasks : (unit -> unit t) Queue.t (* pending tasks *)
-  ; mutable runner : unit k option
+  { mutable state : state
         (* the continuation to resume the runner set by [run]. After it's
            called, the continuation is disposed of and must be set again *)
   ; mutable status : status
@@ -23,21 +27,22 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { tasks = Queue.create (); runner = None; status = Open }
+let create () = { state = Waiting_to_execute (Queue.create ()); status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
   | Open -> (
-    match t.runner with
-    | None ->
-      Queue.push t.tasks f;
+    match t.state with
+    | Stopped () -> assert false
+    | Waiting_to_execute tasks ->
+      Queue.push tasks f;
       k ()
-    | Some r ->
-      assert (Queue.is_empty t.tasks);
-      t.runner <- None;
-      Queue.push t.tasks f;
+    | Waiting_for_tasks r ->
+      let q = Queue.create () in
+      Queue.push q f;
+      t.state <- Waiting_to_execute q;
       resume r () k)
 
 let stop t k =
@@ -45,33 +50,36 @@ let stop t k =
   | Closed -> k ()
   | Open -> (
     t.status <- Closed;
-    match t.runner with
-    | None -> k ()
-    | Some r ->
-      t.runner <- None;
-      resume r () k)
+    match t.state with
+    | Stopped () -> assert false
+    | Waiting_for_tasks r ->
+      t.state <- Stopped ();
+      resume r () k
+    | Waiting_to_execute _ -> k ())
 
 let run t k =
-  match t.runner with
-  | Some _ ->
+  match t.state with
+  | Waiting_for_tasks _ ->
     Code_error.raise "Fiber.Pool.run: concurent calls to run aren't allowed" []
-  | None ->
+  | Stopped () | Waiting_to_execute _ ->
     let n = ref 1 in
     let done_fiber () =
       decr n;
       if !n = 0 then k () else end_of_fiber
     in
     let rec read t =
-      match Queue.pop t.tasks with
+      match
+        Queue.pop
+          (match t.state with
+          | Waiting_for_tasks _ | Stopped () -> assert false
+          | Waiting_to_execute q -> q)
+      with
       | None -> on_finish t
       | Some v ->
         incr n;
         fork (fun () -> v () done_fiber) read_delayed
     and read_delayed () = read t
-    and suspend_k k =
-      assert (t.runner = None);
-      assert (Queue.is_empty t.tasks);
-      t.runner <- Some k
+    and suspend_k k = t.state <- Waiting_for_tasks k
     and on_finish t =
       match t.status with
       | Closed -> done_fiber ()
```

It can be cleaned up a bit, but I fear the code is worse because the reference to `tasks` must re-acquired.

Reply:
Thanks for sharing the code! When you say the reference must be re-acquired, which line do you mean concretely? The `match t.state with` statement in `read`?

Reply:
Yup. That one.

Reply:
I like the data type in this version though. Are you worried about performance impact of reacquiring the reference or the clarity? If performance, could you measure the effect?

Reply:
Btw, all this continuation shuffling turns out to be wrong 😱 

If we do a concurrent run to `run` inside a task, i.e. `Pool.task pool ~f:(fun () -> Pool.run pool)` the check fails. So we should really not be resetting the continuation. I've added a test for this.

Reply:
And given this, I realize that I don't need to set the continuation more than once - it's always the same. So one more mutation gone :)

Reply:
:tada: Sounds great! Did you push the changes?

Reply:
Yes.